### ISSUE TITLE
Target python version based deprecation warnings

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2321]
**Is your feature request related to a problem? Please describe.**
For cases like [PEP 585](https://www.python.org/dev/peps/pep-0585/) where a group of types have been deprecated from the stdlib it can be difficult to keep track of what's been deprecated without stuff falling through the gaps, and without any explicit calls to these types (likely due to them being used in type hints as opposed to directly built) there won't be any runtime deprecation warnings either.

**Describe the solution you'd like**
From an end user perspective this would be solved by Pyright emitting deprecation warnings based on the target python version when it finds references to types (and probably also functions or specific overloads of functions even though that's outside the scope of the presented issue) from the stdlib and possibly also other stubs which have been marked as deprecated.

**Additional context**
This is a follow up for https://github.com/microsoft/pyright/discussions/2300 where some discussion about the nature of how this may work was had.

**Comment by erictraut** [comment]
I've prototyped this functionality, but we've decided not to enable it. We think it will generate too much noise because these types are still used in most Python code today. We are still discussing whether to enable the functionality through a user-controllable setting, but for now we'll just leave it disabled. 

There are also parallel discussions going on in the typing-sig and python-sig forums about creating a standardized way to mark functions and classes as deprecated. If such as standard emerges, we would want to use that within typeshed stubs rather than hard-code knowledge of PEP 585 deprecated classes.
